Add a PoseSelector for use by IDM, MOBIL

jadecastro opened the pull request:
Enables the selection, among several agent cars, a pair of cars that have the closest relative `s`-positions ahead and behind an ego car traveling in the same lane or an adjacent one on a maliput road.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5499)
<!-- Reviewable:end -->

jadecastro commented:
+@liangfok for feature review. Thanks!

liangfok commented:
Completed first pass.

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfHg2lHgMgiX71bpUgD:-KfHg2lHgMgiX71bpUgE:b-cf917v) (no related file):*
BTW, what are the pros / cons of defining a class with no member variables and all static methods vs. a namespace with free methods?

---

*[drake/automotive/pose_selector.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIdGulqSRjQ3DRx7VR:-KfIdGulqSRjQ3DRx7VS:b-ceisuc) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L36)):*
> ```cpp
>       std::numeric_limits<T>::infinity(), 0., 0.));
>   RoadPosition result_trailing = RoadPosition(lane, LanePosition(
>       -std::numeric_limits<T>::infinity(), 0., 0.));
> ```

BTW, in the past I've been defected for using infinity to mean "none" because it could result in major failures if consuming code forgets to check for this special case and decide to use it as an actual value. Can we instead use a safer `enum class` or some other flag to indicate "none"?

---

*[drake/automotive/pose_selector.cc, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfI_jRKisw5VB35X0mG:-KfI_jRKisw5VB35X0mH:b-7nsi83) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L46)):*
> ```cpp
>     if (ego_position.lane->id().id != lane->id().id ||
>         s_agent != ego_position.pos.s) {
>       if (agent_position.lane->id().id == lane->id().id &&
> ```

BTW, consider adding the following check earlier to reject all agent positions that are not in the desired lane:

```
if (agent_position.lane->id().id != lane->id().id) continue;
```

---

*[drake/automotive/pose_selector.cc, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIaRmr0Nf6mWWFxn8m:-KfIaRmr0Nf6mWWFxn8n:bmvtpne) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L47)):*
> ```cpp
>         s_agent != ego_position.pos.s) {
>       if (agent_position.lane->id().id == lane->id().id &&
>           result_trailing.pos.s < s_agent && s_agent < result_leading.pos.s) {
> ```

BTW, consider adding a comment here emphasizing that we purposefully only compare `s` coordinates even when they reside in different lanes.

---

*[drake/automotive/pose_selector.cc, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIS3q4oHwPWg0OG1xu:-KfIS3q5TiPz7p08b4e3:b-x4396) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L70)):*
> ```cpp
>   const maliput::api::GeoPosition& geo_position = maliput::api::GeoPosition(
>       pose.translation().x(), pose.translation().y(), pose.translation().z());
>   return road.ToRoadPosition(geo_position, nullptr, nullptr, nullptr);
> ```

BTW, consider simplifying as follows:

```c++
  return road.ToRoadPosition(
      maliput::api::GeoPosition(pose.translation().x(), pose.translation().y(),
                                pose.translation().z()),
      nullptr, nullptr, nullptr);
```

---

*[drake/automotive/pose_selector.h, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfHc_xGBx7Zk1WyoV5w:-KfHc_xGBx7Zk1WyoV5x:bvqsup9) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L16)):*
> ```objc
> namespace automotive {
> 
> /// PoseSelector contains utilities for selecting among several agent cars those
> ```

BTW, odd grammar: "cars those that".

---

*[drake/automotive/pose_selector.h, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfHfgHCBGJbTK2dlNgN:-KfHfgHCBGJbTK2dlNgO:b2pddv0) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L18)):*
> ```objc
> /// PoseSelector contains utilities for selecting among several agent cars those
> /// that have the closest relative `s`-positions to an ego car traveling in the
> /// same maliput lane or an adjacent one.
> ```

The above sentence is confusing because it states that only relative `s`-positions are used and yet it also states that "adjacent lanes" are considered. How can adjacent lanes be considered without also considering the `r`-coordinates?

---

*[drake/automotive/pose_selector.h, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfI9KSZ1MzN6Ud7vOMu:-KfI9KSZ1MzN6Ud7vOMv:b-jlblja) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L36)):*
> a comparison is made for cars in @p agent_lane.

This sentence could be clarifies by saying "only cars in @p agent_lane are considered".

---

*[drake/automotive/pose_selector.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfI9hPXggtXCbSqBQKG:-KfI9hPXggtXCbSqBQKH:bynvzze) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L38)):*
> infinite distances away from the ego car.

What are the return values? `maliput::api::RoadPosition` objects containing `std::numeric_limits<double>::infinity()`?

---

*[drake/automotive/pose_selector.h, line 41 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIIV0a1KiNoWH5ve4_:-KfIIV0a1KiNoWH5ve4a:b-hj2x8l) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L41)):*
Could this sentence be clarified by stating the following?

> All laterally-adjacent lanes share the same increasing-`s` direction. The curvature of the lanes is small enough such that the increased radius of the outer lanes relative to the inner lanes does not matter.

---

*[drake/automotive/pose_selector.h, line 43 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIJj6TlcGVm_0Db23q:-KfIJj6TlcGVm_0Db23r:bf5c44m) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L43)):*
> ```objc
>   /// `s`-positions of any two cars on the road is meaningful.
>   static const std::pair<maliput::api::RoadPosition, maliput::api::RoadPosition>
>   SelectClosestPositions(const maliput::api::RoadGeometry& road,
> ```

BTW, consider using the verb "Find" instead of "Select" since this method is searching for the closest leading and trailing positions.

---

*[drake/automotive/pose_selector.h, line 45 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIKBiz_TBWGLvRCMAi:-KfIKBiz_TBWGLvRCMAj:b-goaj7h) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L45)):*
> ```objc
>   SelectClosestPositions(const maliput::api::RoadGeometry& road,
>                          const systems::rendering::PoseVector<T>& ego_pose,
>                          const systems::rendering::PoseBundle<T>& agent_poses,
> ```

BTW, in the future when we have more advanced agent planners, isn't it conceivable that agent planners will also make use of this method? If so, consider changing the name of this parameter to be `traffic_poses`.

(Same comment regarding parameter `agent_poses` below.)

---

*[drake/automotive/pose_selector.h, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIUrCRTaBma2Cdg3X1:-KfIUrCRTaBma2Cdg3X2:b-z261pc) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L53)):*
BTW, consider modifying this documentation to reference the above documentation:

> Same as SelectClosestPositions() except it (1) only considers the ego car's lane and (2) only returns the position of the leading vehicle.

---

*[drake/automotive/pose_selector.h, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIXDZn6B2oem6iOetH:-KfIXDZn6B2oem6iOetI:bal0yd5) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L61)):*
> ```objc
>   /// Computes the RoadPosition for a car whose @p pose is located on a given @p
>   /// road.
>   static const maliput::api::RoadPosition CalcRoadPosition(
> ```

BTW, should this be hidden from the public API?

---

*[drake/automotive/test/pose_selector_test.cc, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIakuM86obBj0_KVO2:-KfIakuM86obBj0_KVO3:b-vzibof) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/test/pose_selector_test.cc#L5)):*
> ```cpp
> #include "gtest/gtest.h"
> 
> #include "drake/automotive/maliput/dragway/road_geometry.h"
> ```

BTW, consider also testing with `monolane::ArcLane`.

---

*[drake/automotive/test/pose_selector_test.cc, line 68 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIbMNXCg3HfzP_G1d3:-KfIbMNXCg3HfzP_G1d4:bha68s5) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/test/pose_selector_test.cc#L68)):*
> ```cpp
>   // Lane-coordinate frames are aligned with the x-y world coordinates, with s_i
>   // = 0 for the i-th lane, i ∈ {0, 1}, corresponds to x = 0, and r_i = 0
>   // corresponds to y = (i - 0.5) * kLaneWidth.
> ```

BTW, an ASCII art figure would help me better visualize the dragway that's being used in this test.

---

*[drake/automotive/test/pose_selector_test.cc, line 107 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIcQFlr6Ko-V4PWxFi:-KfIcQFlr6Ko-V4PWxFj:b-qrxke4) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/test/pose_selector_test.cc#L107)):*
> ```cpp
>   EXPECT_EQ(-std::numeric_limits<double>::infinity(), trailing_position.pos.s);
> 
>   // Bump the "just ahead" car into the lane to the left.
> ```

It looks like the "just ahead" vehicle is only being laterally shifted without changing the actual lane it is officially in. Is this correct? If so, shouldn't we officially move it to the left lane and verify the DUT still works?

(Same comment regarding the bumping of the far ahead vehicle to the left lane.)

---

*[drake/automotive/test/pose_selector_test.cc, line 117 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIcn9uFH2nkGF9orpy:-KfIcn9uFH2nkGF9orpz:bx8lv96) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/test/pose_selector_test.cc#L117)):*
> ```cpp
>   EXPECT_EQ(kLeadingSPosition + kSOffset, leading_position.pos.s);
> 
>   // Bump the "just ahead" car into the lane to the left.
> ```

"just ahead" should be "far ahead".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-:-KfIdsLdEvdVMdlcM45n:bl4rucz)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Some comments addressed. 

---

Review status: 0 of 6 files reviewed at latest revision, 18 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfHg2lHgMgiX71bpUgD:-KfM_pIS03qxM7qE5zwt:b-awsnj8) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, what are the pros / cons of defining a class with no member variables and all static methods vs. a namespace with free methods?
</blockquote></details>

Good question. This ultimately prompted me to check the style guide and I think a namespace with free variables is the preferred option, so I've made the appropriate changes. 

Note also that, to avoid having to write out explicit instantiations for each function, I have made everything of type `double` (since the dependencies only work with `double`).

---

*[drake/automotive/pose_selector.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIdGulqSRjQ3DRx7VR:-KfMhk2D5jhWrvWRbn5Y:bujwqk0) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L36)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, in the past I've been defected for using infinity to mean "none" because it could result in major failures if consuming code forgets to check for this special case and decide to use it as an actual value. Can we instead use a safer `enum class` or some other flag to indicate "none"?
</blockquote></details>

In this case, I don't think we really want a `none`, because we still need to provide the caller with valid distance measure (effectively, a distance at positive or negative infinity). So, consumer code should be prepared to consume either a finite `T` value or an infinite value (this is stated as a contract in the function doxygen). Is your case different?  

Do you have an alternative to `std::numeric_limits<double>::infinity()`? I can see using some arbitrarily large number `kHugeDistance`, but I suspect this is beside your original point?

---

*[drake/automotive/pose_selector.cc, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfI_jRKisw5VB35X0mG:-KfMmIIDfjbh6d3RXlcD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L46)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding the following check earlier to reject all agent positions that are not in the desired lane:

```
if (agent_position.lane->id().id != lane->id().id) continue;
```
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.cc, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIaRmr0Nf6mWWFxn8m:-KfMoeybj8Mcj1B70pMI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L47)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a comment here emphasizing that we purposefully only compare `s` coordinates even when they reside in different lanes.
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.cc, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIS3q4oHwPWg0OG1xu:-KfMpeXK2iRKCLUZ42VA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L70)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider simplifying as follows:

```c++
  return road.ToRoadPosition(
      maliput::api::GeoPosition(pose.translation().x(), pose.translation().y(),
                                pose.translation().z()),
      nullptr, nullptr, nullptr);
```
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfHc_xGBx7Zk1WyoV5w:-KfMprhsNUVFyGtGvmAt:bvcgppk) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L16)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, odd grammar: "cars those that".
</blockquote></details>

I have removed this comment block.

---

*[drake/automotive/pose_selector.h, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfHfgHCBGJbTK2dlNgN:-KfMqGCStDsMh1FRArNO:bvcgppk) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

The above sentence is confusing because it states that only relative `s`-positions are used and yet it also states that "adjacent lanes" are considered. How can adjacent lanes be considered without also considering the `r`-coordinates?
</blockquote></details>

I have removed this comment block.

---

*[drake/automotive/pose_selector.h, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfI9KSZ1MzN6Ud7vOMu:-KfNeTmEeiE3NYhaGLjO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L36)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

> a comparison is made for cars in @p agent_lane.

This sentence could be clarifies by saying "only cars in @p agent_lane are considered".
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfI9hPXggtXCbSqBQKG:-KfNtI3Doa9W7Giao8t4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L38)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

> infinite distances away from the ego car.

What are the return values? `maliput::api::RoadPosition` objects containing `std::numeric_limits<double>::infinity()`?
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 43 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIJj6TlcGVm_0Db23q:-KfNuI_SfW0Ogvafyb-Y:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L43)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider using the verb "Find" instead of "Select" since this method is searching for the closest leading and trailing positions.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-:-KfNucnlmAAcf6UvCBHm:b-hb0k8w)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 3 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/automotive/pose_selector.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIdGulqSRjQ3DRx7VR:-KfORvEvY-LqwwvkryZJ:bxadgl) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L36)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

In this case, I don't think we really want a `none`, because we still need to provide the caller with valid distance measure (effectively, a distance at positive or negative infinity). So, consumer code should be prepared to consume either a finite `T` value or an infinite value (this is stated as a contract in the function doxygen). Is your case different?  

Do you have an alternative to `std::numeric_limits<double>::infinity()`? I can see using some arbitrarily large number `kHugeDistance`, but I suspect this is beside your original point?
</blockquote></details>

Here's my situation from the past:

https://reviewable.io/reviews/robotlocomotion/drake/4597#-K_9Qv9wHKuqH3TpEdeZ

It was in the context of the depth sensor, and a value of negative infinity was being used to indicate "the object is too close to the sensor for a valid depth measurement to be made". I changed it to return zero instead based on reviewer comments.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/pose_selector.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIdGulqSRjQ3DRx7VR:-KfOVWZxmWiTUSE2VEIh:bw8ga1y) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L36)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Here's my situation from the past:

https://reviewable.io/reviews/robotlocomotion/drake/4597#-K_9Qv9wHKuqH3TpEdeZ

It was in the context of the depth sensor, and a value of negative infinity was being used to indicate "the object is too close to the sensor for a valid depth measurement to be made". I changed it to return zero instead based on reviewer comments.
</blockquote></details>

But in this PR, negative infinity actually makes sense, because its a measurement how far behind us the trailing car is, for which "as far as infinity" makes sense.  That's not the case for a depth camera forward raycast.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/automotive/pose_selector.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIdGulqSRjQ3DRx7VR:-KfOX8qtd6xcINlcrzLP:b-e0h0yg) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L36)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

But in this PR, negative infinity actually makes sense, because its a measurement how far behind us the trailing car is, for which "as far as infinity" makes sense.  That's not the case for a depth camera forward raycast.
</blockquote></details>

The level of risk depends on how we anticipate the return values being used. For example, problems may arise if the user attempts to average an infinity value with a non-infinity value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/automotive/pose_selector.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIdGulqSRjQ3DRx7VR:-KfQyDM138P3E0uOLqaA:b-9d2cz6) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L36)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

The level of risk depends on how we anticipate the return values being used. For example, problems may arise if the user attempts to average an infinity value with a non-infinity value.
</blockquote></details>

A thought just occurred to me. Oftentimes when I drive, I not only look at the vehicle immediately in front of me but I look at the vehicles ahead of that one. Thus, maybe the return value of this method should be two vectors: a sorted list of poses of vehicles in front of the ego car and a sorted list of poses of vehicles behind the ego car. Thus, an empty vector would represent the "none" scenario.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/automotive/pose_selector.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIdGulqSRjQ3DRx7VR:-KfR-Gki3mSxtxW9P8E7:bua6xk0) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L36)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

A thought just occurred to me. Oftentimes when I drive, I not only look at the vehicle immediately in front of me but I look at the vehicles ahead of that one. Thus, maybe the return value of this method should be two vectors: a sorted list of poses of vehicles in front of the ego car and a sorted list of poses of vehicles behind the ego car. Thus, an empty vector would represent the "none" scenario.
</blockquote></details>

This is a bit of feature creep for this PR, I think? I totally agree, but for the purposes of IDM and MOBIL, only the nearest cars count. 

---

*[drake/automotive/pose_selector.h, line 45 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIKBiz_TBWGLvRCMAi:-KfOmkMTKjUje1fKGv7t:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L45)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, in the future when we have more advanced agent planners, isn't it conceivable that agent planners will also make use of this method? If so, consider changing the name of this parameter to be `traffic_poses`.

(Same comment regarding parameter `agent_poses` below.)
</blockquote></details>

Done.

---

*[drake/automotive/test/pose_selector_test.cc, line 117 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIcn9uFH2nkGF9orpy:-KfQUDP94QgMJHaFfHvO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/test/pose_selector_test.cc#L117)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

"just ahead" should be "far ahead".
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/automotive/pose_selector.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIdGulqSRjQ3DRx7VR:-KfRDkDSzcpq10G_9yfe:bk9jvz6) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.cc#L36)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

This is a bit of feature creep for this PR, I think? I totally agree, but for the purposes of IDM and MOBIL, only the nearest cars count. 
</blockquote></details>

Yeah, I only mention it because it has relevance to the question of how to represent "none".

---

*[drake/automotive/pose_selector.cc, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfREIZTw_HnZuiqXJBJ:-KfREIZTw_HnZuiqXJBK:bksc503) ([raw file](https://github.com/robotlocomotion/drake/blob/93c783be0d7a1c780ad807ec890a72fac1587aaa/drake/automotive/pose_selector.cc#L24)):*
> ```cpp
>   const RoadPosition& ego_position =
>       CalcRoadPosition(road, ego_pose.get_isometry());
>   std::cerr << " ego_position.pos.s: " << ego_position.pos.s << std::endl;
> ```

BTW, should this be removed?

---

*[drake/automotive/test/pose_selector_test.cc, line 117 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIcn9uFH2nkGF9orpy:-KfRDwEXKWG9lM6qxsTy:b-wcz65n) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/test/pose_selector_test.cc#L117)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

No change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 2 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/automotive/pose_selector.cc, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfREIZTw_HnZuiqXJBJ:-KfRHHjQYV_M7b_7ZnbE:brx6qxr) ([raw file](https://github.com/robotlocomotion/drake/blob/93c783be0d7a1c780ad807ec890a72fac1587aaa/drake/automotive/pose_selector.cc#L24)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, should this be removed?
</blockquote></details>

Sorry; I hit Publish to respond to another comment along with drafts and that prompted me to push what I had, which unfortunately had some warts ;)  Removed.

---

*[drake/automotive/pose_selector.h, line 41 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIIV0a1KiNoWH5ve4_:-KfR3KrkF69NbR5jfuru:b1lko2n) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L41)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Could this sentence be clarified by stating the following?

> All laterally-adjacent lanes share the same increasing-`s` direction. The curvature of the lanes is small enough such that the increased radius of the outer lanes relative to the inner lanes does not matter.
</blockquote></details>

Good point. For now, I am restricting to straight multi-lane roads. I've added a note to that effect.

---

*[drake/automotive/pose_selector.h, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIUrCRTaBma2Cdg3X1:-KfRJf856CRxXoTfd-Uq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L53)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider modifying this documentation to reference the above documentation:

> Same as SelectClosestPositions() except it (1) only considers the ego car's lane and (2) only returns the position of the leading vehicle.
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIXDZn6B2oem6iOetH:-KfRGUzZ8NWxHgt0u9Re:b-ruowim) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/pose_selector.h#L61)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, should this be hidden from the public API?
</blockquote></details>

It will be called directly in IDM and MOBIL. Anyhow, this is no longer a class.

---

*[drake/automotive/test/pose_selector_test.cc, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIakuM86obBj0_KVO2:-KfRHCTPattWmSku6pR2:btj0uoi) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/test/pose_selector_test.cc#L5)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider also testing with `monolane::ArcLane`.
</blockquote></details>

I've clarified that this is only useful for straight lanes (see comment above). 

---

*[drake/automotive/test/pose_selector_test.cc, line 68 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIbMNXCg3HfzP_G1d3:-KfRInEltQlK1QJbcYGr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/test/pose_selector_test.cc#L68)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, an ASCII art figure would help me better visualize the dragway that's being used in this test.
</blockquote></details>

Done.

---

*[drake/automotive/test/pose_selector_test.cc, line 107 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIcQFlr6Ko-V4PWxFi:-KfRJ-DfJDWmOdIRzf4a:b8f9mkb) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/test/pose_selector_test.cc#L107)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

It looks like the "just ahead" vehicle is only being laterally shifted without changing the actual lane it is officially in. Is this correct? If so, shouldn't we officially move it to the left lane and verify the DUT still works?

(Same comment regarding the bumping of the far ahead vehicle to the left lane.)
</blockquote></details>

`PoseBundle` doesn't take lane information, so `traffic_poses` contains no notion of lane.

---

*[drake/automotive/test/pose_selector_test.cc, line 117 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfIcn9uFH2nkGF9orpy:-KfRJUoTIUvrG-QDxeSB:bdv43q9) ([raw file](https://github.com/robotlocomotion/drake/blob/09736f66adba3fe854de93ba4c15f1684e6f0786/drake/automotive/test/pose_selector_test.cc#L117)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

No change?
</blockquote></details>

Sorry! I applied my fix in the wrong place. Done. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-:-KfRWVO4TSWkpDxJ2O0m:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
+@jwnimmer-tri for platform, automotive review, please.

jwnimmer-tri commented:
:car: approved
-@jwnimmer-tri +@ggould-tri for platform review, please.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-:-KfRhqzw4eSANHLmXhqg:b-jw546z)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
This is really an excellently clear interface and test.  Thanks!
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 3 files at r2, 3 of 4 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/pose_selector.cc, line 35 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfRq61TJ9s_7z2BP5hK:-KfRq61TJ9s_7z2BP5hL:bitc4l2) ([raw file](https://github.com/robotlocomotion/drake/blob/145733d9ffb019017bafa61c530b1a1ad198265f/drake/automotive/pose_selector.cc#L35)):*
> ```cpp
>   RoadPosition result_trailing = RoadPosition(
>       lane, LanePosition(-std::numeric_limits<double>::infinity(), 0., 0.));
>   for (int i = 0; i < traffic_poses.get_num_poses(); ++i) {
> ```

BTW -- I am sad that this cannot be a range-based for loop.  Someone should consider implementing `PoseBundle::begin` and `PoseBundle::end` to enable that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-:-KfRrzYF_Y63jWlbdki-:b-7chj31)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[drake/automotive/pose_selector.cc, line 35 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5499#-KfRq61TJ9s_7z2BP5hK:-KfSKbGhly7qwmGsPHl_:bturhq) ([raw file](https://github.com/robotlocomotion/drake/blob/145733d9ffb019017bafa61c530b1a1ad198265f/drake/automotive/pose_selector.cc#L35)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW -- I am sad that this cannot be a range-based for loop.  Someone should consider implementing `PoseBundle::begin` and `PoseBundle::end` to enable that.
</blockquote></details>

I agree, and would be willing to work on that feature, since it would clearly benefit most all use cases for `PoseBundle`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5499)*
<!-- Sent from Reviewable.io -->
